crates/ruff_python_semantic/src/model.rs

```diff
@@ -1575,6 +1575,31 @@ impl<'a> SemanticModel<'a> {
             .intersects(SemanticModelFlags::DEFERRED_TYPE_DEFINITION)
     }
 
+    pub const fn in_valid_context_for_deferred_type_definition(&self) -> bool {
+        if self.in_typing_literal() {
+            return false;
+        }
+        if !self.in_type_definition() {
+            return false;
+        }
+        // `in_deferred_type_definition()` will only be `true` if we're now visiting the deferred nodes
+        // after having already traversed the source tree once. If we're now visiting the deferred nodes,
+        // we can't defer again, or we'll infinitely recurse!
+        if self.in_deferred_type_definition() {
+            return false;
+        }
+        if self.in_type_checking_block() || self.flags.intersects(SemanticModelFlags::STUB_FILE) {
+            return true;
+        }
+        if !self
+            .flags
+            .intersects(SemanticModelFlags::FUTURE_ANNOTATIONS)
```

Comment:
looks like there isn't a `self.future_annotations()`, maybe there should be?

Reply:
I worry that it's a *bit* of a footgun, as there aren't really very many situations where you want to be able to differentiate between `.py` files with `from __future__ import annotations` and stub files, and it's pretty easy to forget that you need to use `self.future_annotations_or_stub()` rather than `self.future_annotations()`. It's for those reasons that I... removed our previous `self.future_annotations()` method in https://github.com/astral-sh/ruff/commit/1dc93107dce911a535bdd9ac0d918f8052b7e326 😄 but, I'll add it back